File: android-control/app/src/test/java/cn/xiaoli/control/ServerUrlPolicyTest.java
package cn.xiaoli.control;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertThrows;
import static org.junit.Assert.assertTrue;

import org.junit.Test;

public final class ServerUrlPolicyTest {
    @Test
    public void acceptsHttpsAddressAndBuildsCanonicalDashboardUrl() {
        ServerUrlPolicy policy = ServerUrlPolicy.fromUserInput(
                "  https://Bridge.Example.com:443/previous/path?query=yes#fragment  ");

        assertEquals("https://bridge.example.com", policy.origin());
        assertEquals("https://bridge.example.com/mobile/", policy.dashboardUrl());
    }

    @Test
    public void acceptsHttpsAddressWithNonDefaultPort() {
        ServerUrlPolicy policy = ServerUrlPolicy.fromUserInput("https://bridge.example.com:8443");

        assertEquals("https://bridge.example.com:8443", policy.origin());
        assertEquals("https://bridge.example.com:8443/mobile/", policy.dashboardUrl());
    }

    @Test
    public void rejectsBlankAndNonHttpsBridgeAddresses() {
        assertThrows(IllegalArgumentException.class, () -> ServerUrlPolicy.fromUserInput(""));
        assertThrows(IllegalArgumentException.class, () -> ServerUrlPolicy.fromUserInput("http://bridge.example.com"));
        assertThrows(IllegalArgumentException.class, () -> ServerUrlPolicy.fromUserInput("file:///tmp/mobile.html"));
        assertThrows(IllegalArgumentException.class, () -> ServerUrlPolicy.fromUserInput("content://cn.xiaoli.control/mobile"));
        assertThrows(IllegalArgumentException.class, () -> ServerUrlPolicy.fromUserInput("javascript:alert(1)"));
    }

    @Test
    public void allowsOnlyHttpsNavigationOnTheExactConfiguredOrigin() {
        ServerUrlPolicy policy = ServerUrlPolicy.fromUserInput("https://bridge.example.com");

        assertTrue(policy.isAllowedNavigation("https://bridge.example.com/mobile/"));
        assertTrue(policy.isAllowedNavigation("https://BRIDGE.EXAMPLE.COM/api/mobile/v1/status?fresh=1"));
        assertTrue(policy.isAllowedNavigation("https://bridge.example.com:443/mobile/app.js"));
        assertFalse(policy.isAllowedNavigation("http://bridge.example.com/mobile/"));
        assertFalse(policy.isAllowedNavigation("https://bridge.example.com:8443/mobile/"));
        assertFalse(policy.isAllowedNavigation("file:///android_asset/index.html"));
        assertFalse(policy.isAllowedNavigation("content://cn.xiaoli.control/mobile"));
        assertFalse(policy.isAllowedNavigation("javascript:alert(1)"));
    }

    @Test
    public void rejectsHostConfusionAndCredentials() {
        ServerUrlPolicy policy = ServerUrlPolicy.fromUserInput("https://bridge.example.com");

        assertFalse(policy.isAllowedNavigation("https://bridge.example.com.evil.test/mobile/"));
        assertFalse(policy.isAllowedNavigation("https://bridge.example.com@evil.test/mobile/"));
        assertFalse(policy.isAllowedNavigation("https://evil.test/?next=https://bridge.example.com"));
        assertThrows(
                IllegalArgumentException.class,
                () -> ServerUrlPolicy.fromUserInput("https://user:password@bridge.example.com"));
    }

    @Test
    public void malformedNavigationIsRejectedWithoutThrowing() {
        ServerUrlPolicy policy = ServerUrlPolicy.fromUserInput("https://bridge.example.com");

        assertFalse(policy.isAllowedNavigation(null));
        assertFalse(policy.isAllowedNavigation(""));
        assertFalse(policy.isAllowedNavigation("not a url"));
    }
}

File: ios-control/tests/serverUrlPolicy.test.ts
import assert from 'node:assert/strict';
import test from 'node:test';

import { createServerUrlPolicy, isApkDownload } from '../src/serverUrlPolicy';

test('canonicalizes an HTTPS origin and discards path, query, and fragment', () => {
  const policy = createServerUrlPolicy('  HTTPS://Example.COM:443/setup?q=1#part  ');
  assert.equal(policy.origin, 'https://example.com');
  assert.equal(policy.dashboardUrl, 'https://example.com/mobile/');
  assert.equal(policy.host, 'example.com');
  assert.equal(policy.effectivePort, 443);
});

test('keeps a non-default port', () => {
  const policy = createServerUrlPolicy('https://Bridge.Example:8443/path');
  assert.equal(policy.origin, 'https://bridge.example:8443');
  assert.equal(policy.dashboardUrl, 'https://bridge.example:8443/mobile/');
});

test('allows only the exact HTTPS origin', () => {
  const policy = createServerUrlPolicy('https://bridge.example');
  assert.equal(policy.isAllowedNavigation('https://bridge.example/mobile/'), true);
  assert.equal(policy.isAllowedNavigation('https://BRIDGE.example:443/api/mobile/v1/status'), true);
  assert.equal(policy.isAllowedNavigation('https://bridge.example.evil.test/mobile/'), false);
  assert.equal(policy.isAllowedNavigation('https://evil.test/?next=https://bridge.example'), false);
  assert.equal(policy.isAllowedNavigation('http://bridge.example/mobile/'), false);
  assert.equal(policy.isAllowedNavigation('https://bridge.example:8443/mobile/'), false);
  assert.equal(policy.isAllowedNavigation('https://user@bridge.example/mobile/'), false);
});

test('rejects invalid or unsafe configuration addresses', () => {
  const invalid = [
    '',
    'example.com',
    'http://example.com',
    'file:///tmp/index.html',
    'content://settings',
    'javascript:alert(1)',
    'https://user:pass@example.com',
    'https://@example.com',
    'https://example.com:',
    'https://example.com:0',
    'https://example.com:65536',
    'https:\\example.com',
    'https://example.com\u0000.evil.test',
  ];

  for (const candidate of invalid) {
    assert.throws(() => createServerUrlPolicy(candidate), /valid HTTPS/, candidate);
  }
});

test('malformed navigation candidates return false instead of throwing', () => {
  const policy = createServerUrlPolicy('https://bridge.example');
  for (const candidate of ['', 'not a URL', 'data:text/html,hello', 'https://bridge.example:']) {
    assert.doesNotThrow(() => policy.isAllowedNavigation(candidate));
    assert.equal(policy.isAllowedNavigation(candidate), false);
  }
});

test('recognizes APK downloads by parsed path', () => {
  assert.equal(isApkDownload('https://bridge.example/downloads/xiaoli-control.apk'), true);
  assert.equal(isApkDownload('https://bridge.example/downloads/APP.APK?version=2'), true);
  assert.equal(isApkDownload('https://bridge.example/mobile/'), false);
  assert.equal(isApkDownload('not a URL'), false);
});
